File: 2-oop/oop-examples/cats.py
class Animal:
    def __init__(self, name, age):
        self.__name = name
        self.__age = age

    @property
    def name(self):
        return self.__name

    @name.setter
    def name(self, value):
        self.__name = value

    @property
    def age(self):
        return self.__age

    @age.setter
    def age(self, value):
        self.__age = value

    def speak(self):
        raise NotImplementedError("Этот метод должен быть переопределён в наследнике.")

class Cat(Animal):
    def __init__(self, name, age):
        super().__init__(name, age)

    def speak(self):
        print(f"{self.name} мяукает!")

    def purr(self):
        print(f"{self.name} мурлычет.")


if __name__ == "__main__":
    cat = Cat("Барсик", 3)
    print("Имя кота:", cat.name)
    print("Возраст кота:", cat.age)

    cat.speak()
    cat.purr()

    cat.age = 4
    cat.name = "Мурзик"

    print("\nПосле изменения данных через сеттеры:")
    print("Новое имя кота:", cat.name)
    print("Новый возраст кота:", cat.age)

File: 2-oop/oop-examples/cats.cs
using System;

// Абстрактный класс Animal
public abstract class Animal
{
    private string _name;
    private int _age;

    public Animal(string name, int age)
    {
        _name = name;
        _age = age;
    }

    // Свойства (property) для имени
    public string Name
    {
        get { return _name; }
        set { _name = value; }
    }

    // Свойства (property) для возраста
    public int Age
    {
        get { return _age; }
        set { _age = value; }
    }

    // Абстрактный метод - будет переопределён в наследнике
    public abstract void Speak();
}

// Класс Cat наследуется от Animal
public class Cat : Animal
{
    public Cat(string name, int age) : base(name, age) {}

    public override void Speak()
    {
        Console.WriteLine($"{Name} мяукает!");
    }

    public void Purr()
    {
        Console.WriteLine($"{Name} мурлычет.");
    }
}

public class Program
{
    public static void Main()
    {
        Cat cat = new Cat("Барсик", 3);

        Console.WriteLine("Имя кота: " + cat.Name);
        Console.WriteLine("Возраст кота: " + cat.Age);

        cat.Speak();
        cat.Purr();

        cat.Age = 4;
        cat.Name = "Мурзик";

        Console.WriteLine("\nПосле изменения данных через сеттеры:");
        Console.WriteLine("Новое имя кота: " + cat.Name);
        Console.WriteLine("Новый возраст кота: " + cat.Age);
    }
}
